internal/runbits/buildlogfile/buildlogfile.go

```diff
@@ -0,0 +1,245 @@
+package buildlogfile
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/ActiveState/cli/internal/constants"
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/output"
+	"github.com/ActiveState/cli/pkg/platform/model"
+	"github.com/ActiveState/cli/pkg/platform/runtime/artifact"
+	"github.com/ActiveState/cli/pkg/platform/runtime/setup/buildlog"
+)
+
+// maxConcurrency is the maximum number of artifact-build-logs that we download in parallel in order to add the information to the build log file.
+const maxConcurrency = 5
+
+type BuildLogFile struct {
+	out              output.Outputer
+	logFile          *os.File
+	numInstalled     int
+	numToBeInstalled int
+	ctx              context.Context
+	cancel           context.CancelFunc
+	logsCh           chan artifactLog
+}
+
+type artifactLog struct {
+	artifactID     artifact.ArtifactID
+	artifactName   string
+	unsignedLogURI string
+}
+
+// verboseLogging returns true if the user provided an environment variable for it
+func verboseLogging() bool {
+	return os.Getenv(constants.LogBuildVerboseEnvVarName) == "true"
+}
+
+// downloadAndAddArtifactLog downloads an artifact build log and adds it to the build log file
+func (bl *BuildLogFile) downloadAndAddArtifactLog(ul artifactLog, ctx context.Context) error {
+	unsignedURL, err := url.Parse(ul.unsignedLogURI)
+	if err != nil {
+		return errs.Wrap(err, "Could not parse log URL %s", ul.unsignedLogURI)
+	}
+	logURL, err := model.SignS3URL(unsignedURL)
+	if err != nil {
+		return errs.Wrap(err, "Could not sign log url %s", unsignedURL)
+	}
+
+	// download the log and stream it line-by-line
+	logging.Debug("downloading logURI: %s", logURL.String())
+	req, err := http.NewRequestWithContext(ctx, "GET", logURL.String(), nil)
+	if err != nil {
+		return errs.Wrap(err, "Failed to create GET request for logURL.")
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return errs.Wrap(err, "Failed to execute HTTP GET request for logURL.")
+	}
+
+	defer resp.Body.Close()
+	scanner := bufio.NewScanner(resp.Body)
+	for scanner.Scan() {
+		line := scanner.Bytes()
+		// we need to unmarshal every line
+		var am buildlog.ArtifactProgressMessage
+		if err := json.Unmarshal(line, &am); err != nil {
+			return errs.Wrap(err, "Failed to unmarshal build log line")
+		}
+		if err := bl.BuildArtifactProgress(ul.artifactID, ul.artifactName, am.Timestamp, am.Body.Message, am.Body.Facility, am.PipeName, am.Source); err != nil {
+			return errs.Wrap(err, "Failed to write build log line")
+		}
+	}
+	return nil
+}
+
+func New(out output.Outputer) (*BuildLogFile, error) {
+	logFile, err := os.CreateTemp("", fmt.Sprintf("build-log-%s", time.Now().Format("060102030405")))
```

Comment:
Can we use unix timestamp instead? This is needlessly vague.

Reply:
I added a human-readable timestamp.  The `CreateTemp` function adds a unique string (that is timestamp based) to the file anyways. If our only concern is uniqueness, it is probably better to *just* rely on that mechanism, as in the current commit.